pkg/cloudprovider/providers/azure/azure_managedDiskController.go

```diff
@@ -35,56 +41,108 @@ type ManagedDiskController struct {
+	AvailabilityZones  string
+	Tags               map[string]string
+	StorageAccountType storage.SkuName
+}
+
 func newManagedDiskController(common *controllerCommon) (*ManagedDiskController, error) {
 	return &ManagedDiskController{common: common}, nil
 }
 
 //CreateManagedDisk : create managed disk
-func (c *ManagedDiskController) CreateManagedDisk(diskName string, storageAccountType storage.SkuName, resourceGroup string,
-	sizeGB int, tags map[string]string) (string, error) {
-	glog.V(4).Infof("azureDisk - creating new managed Name:%s StorageAccountType:%s Size:%v", diskName, storageAccountType, sizeGB)
+func (c *ManagedDiskController) CreateManagedDisk(options *ManagedDiskOptions) (string, error) {
+	var zones sets.String
+	var activeZones sets.String
+	var err error
+	glog.V(4).Infof("azureDisk - creating new managed Name:%s StorageAccountType:%s Size:%v", options.DiskName, options.StorageAccountType, options.SizeGB)
+
+	// Get active zones which have nodes running on.
+	activeZones, err = c.common.cloud.GetActiveZones()
+	if err != nil {
+		return "", fmt.Errorf("error querying active zones: %v", err)
+	}
+
+	// Validate and choose availability zone for creating disk.
+	if options.Zoned && !options.ZonePresent && !options.ZonesPresent {
+		// Neither "zone" or "zones" specified. Pick a zone randomly selected
+		// from all active zones where Kubernetes cluster has a node.
+		zones = activeZones
+	} else if !options.ZonePresent && options.ZonesPresent {
+		// Choose zone from specified zones.
+		if zones, err = util.ZonesToSet(options.AvailabilityZones); err != nil {
+			return "", err
+		}
+	} else if options.ZonePresent && !options.ZonesPresent {
+		if err := util.ValidateZone(options.AvailabilityZone); err != nil {
+			return "", err
+		}
+		zones = make(sets.String)
+		zones.Insert(options.AvailabilityZone)
+	}
 
+	// insert original tags to newTags
 	newTags := make(map[string]*string)
 	azureDDTag := "kubernetes-azure-dd"
 	newTags["created-by"] = &azureDDTag
-
-	// insert original tags to newTags
-	if tags != nil {
-		for k, v := range tags {
+	if options.Tags != nil {
+		for k, v := range options.Tags {
 			// Azure won't allow / (forward slash) in tags
 			newKey := strings.Replace(k, "/", "-", -1)
 			newValue := strings.Replace(v, "/", "-", -1)
 			newTags[newKey] = &newValue
 		}
 	}
 
-	diskSizeGB := int32(sizeGB)
+	diskSizeGB := int32(options.SizeGB)
 	model := compute.Disk{
 		Location: &c.common.location,
 		Tags:     newTags,
 		Sku: &compute.DiskSku{
-			Name: compute.StorageAccountTypes(storageAccountType),
+			Name: compute.StorageAccountTypes(options.StorageAccountType),
 		},
 		DiskProperties: &compute.DiskProperties{
 			DiskSizeGB:   &diskSizeGB,
 			CreationData: &compute.CreationData{CreateOption: compute.Empty},
-		}}
+		},
+	}
+
+	if options.ResourceGroup == "" {
+		options.ResourceGroup = c.common.resourceGroup
+	}
+	if len(zones.List()) > 0 {
+		createAZ := util.ChooseZoneForVolume(zones, options.PVCName)
```

Comment:
this validation should be in the logic, https://github.com/kubernetes/kubernetes/pull/66553/files#diff-db9a5ad5d2cc7740ca2b73ad9d904fa2R77 zone selection + validation should be early on as possible

Reply:
ACK, will move to above zone selection part.